extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -59,34 +70,73 @@ private static QuarkusSecurityIdentity augmentIdentity(SecurityIdentity security
         if (sessionCookie != null) {
             String[] tokens = sessionCookie.getValue().split(COOKIE_DELIM);
-            return authenticate(identityProviderManager, new IdTokenCredential(tokens[0], context))
-                    .thenCompose(new Function<SecurityIdentity, CompletionStage<SecurityIdentity>>() {
+            String idToken = tokens[0];
+            String accessToken = tokens[1];
+            String refreshToken = tokens[2];
+
+            return authenticate(identityProviderManager, new IdTokenCredential(idToken, context))
+                    .handle(new BiFunction<SecurityIdentity, Throwable, SecurityIdentity>() {
                         @Override
-                        public CompletionStage<SecurityIdentity> apply(SecurityIdentity securityIdentity) {
-                            return CompletableFuture
-                                    .completedFuture(augmentIdentity(securityIdentity, tokens[1], tokens[2], context));
+                        public SecurityIdentity apply(SecurityIdentity identity, Throwable throwable) {
+                            if (isLogout(context, configContext)) {
+                                // force logout, the challenge will be built accordingly and redirect to the logout endpoint at the OP
+                                throw new AuthenticationFailedException();
+                            }
+
+                            // if authentication failed, the reason could be the token no longer being valid, so we try a 
+                            // silent refresh if required
+                            if (throwable != null) {
+                                if ((identity = trySilentRefresh(configContext, idToken, refreshToken, context)) == null) {
+                                    // if the refresh fails, we just propagate the original exception
+                                    if (throwable instanceof AuthenticationFailedException) {
+                                        throw AuthenticationFailedException.class.cast(throwable);
+                                    }
+                                    throw new AuthenticationFailedException(throwable);
+                                }
+                            }
+
+                            return augmentIdentity(identity, accessToken, refreshToken, context);
                         }
                     });
         }
 
         // start a new session by starting the code flow dance
-        return performCodeFlow(identityProviderManager, context, resolver);
+        return performCodeFlow(identityProviderManager, context, configContext);
     }
 
     public CompletionStage<ChallengeData> getChallenge(RoutingContext context, DefaultTenantConfigResolver resolver) {
         TenantConfigContext configContext = resolver.resolve(context, false);
+        ServerCookie sessionCookie = (ServerCookie) context.cookieMap().get(SESSION_COOKIE_NAME);
```

Comment:
We can probably just use `removeCookie` here because it is then removed anyway if it is logout or not

Reply:
The problem is that the your last changes are setting the value of the cookie to an empty string. So, removing here would make hard to get the value and handle it accordingly.

Reply:
@pedroigor Can we then update `removeCookie` to return the value which was reset ?

Reply:
Sure, if you think is better. For me, the way it is OK too.

Reply:
@pedroigor we can have 1 remove cookie call which returns the cookie value instead of 2 calls. simple to do so why not :-)